src/MongoDB.Driver/Linq/Linq3Implementation/Ast/PipelineOptimizer/AstGroupPipelineOptimizer.cs

```diff
@@ -0,0 +1,413 @@
+                    if (_renderedAccumulatorExpressions[i].Equals(renderedAccumulatorExpression))
+                    {
+                        return _accumulatorFields[i].Path;
+                    }
+                }
+
+                var accumulatorFieldName = $"__agg{_accumulatorCounter++}";
+                var accumulatorField = AstExpression.AccumulatorField(accumulatorFieldName, value);
+                _accumulatorFields.Add(accumulatorField);
+                _renderedAccumulatorExpressions.Add(renderedAccumulatorExpression);
+                return accumulatorFieldName;
+            }
+
+            public IEnumerator<AstAccumulatorField> GetEnumerator() => _accumulatorFields.GetEnumerator();
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        }
+
+        private class AccumulatorMover : AstNodeVisitor
+        {
+            #region static
+            public static TNode MoveAccumulators<TNode>(AccumulatorSet accumulators, TNode node)
+                where TNode : AstNode
+            {
+                var mover = new AccumulatorMover(accumulators);
+                return mover.VisitAndConvert(node);
+            }
+            #endregion
+
+            private readonly AccumulatorSet _accumulators;
+
+            private AccumulatorMover(AccumulatorSet accumulator)
+            {
+                _accumulators = accumulator;
+            }
+
+            public override AstNode VisitFilterField(AstFilterField node)
+            {
+                // "_elements.0.X" => { __agg0 : { $first : "$$ROOT" } } + "__agg0.X"
+                if (node.Path.StartsWith("_elements.0."))
+                {
+                    var accumulatorExpression = AstExpression.AccumulatorExpression(AstAccumulatorOperator.First, AstExpression.Var("ROOT"));
+                    var accumulatorFieldName = _accumulators.AddAccumulatorExpression(accumulatorExpression);
+                    var restOfPath = node.Path.Substring("_elements.0.".Length);
+                    var rewrittenPath = $"{accumulatorFieldName}.{restOfPath}";
+                    return AstFilter.Field(rewrittenPath, node.Serializer);
+                }
+
+                if (node.Path == "_elements" || node.Path.StartsWith("_elements."))
+                {
+                    throw new UnableToRemoveReferenceToElementsException();
+                }
+
+                return base.VisitFilterField(node);
+            }
+
+            public override AstNode VisitGetFieldExpression(AstGetFieldExpression node)
+            {
+                if (node.FieldName is AstConstantExpression constantFieldName &&
+                    constantFieldName.Value.IsString &&
+                    constantFieldName.Value.AsString == "_elements" &&
+                    node.Input is AstVarExpression varExpression &&
+                    varExpression.Name == "ROOT")
+                {
+                    throw new UnableToRemoveReferenceToElementsException();
```

Comment:
We're only going to reach here if we weren't able to optimize away the reference to `_elements`.